### ISSUE TITLE
OmniSR model occasionally produces black image during upscaling

### ISSUE THREAD

**Issue by pokepress** [neosr-project/neosr#96]
I've been training an omnisr model, and noticed that occasionally a frame from the video I'm upscaling will have a completely black image. It's extremely peculiar in that:

- The frames that turn out black vary based on the iteration of the model that was used.
- There tend to be fewer black frames as the model is trained more, but I'm still having issues.
- The issue is incredibly specific to the content of the frame. For example, the frame will sometimes render properly if I save it as a jpeg (which is lossy, of course) and upscale it individually, but not if I save the frame as a PNG or TIFF (lossless, obviously). Sometimes a few frames in a row will produce a few black frames in a row if the frames are very similar.
- The issue only seems to occur on the newer "TPCi" era episodes I've been working with, not the older "4Kids" episodes. The episodes obviously used a different processing chain, but notably the the "4Kids" episodes tend to have more fake film grain applied (the series went to digital production late in season 5). Both eras are in the training data.

I'm guessing this is either some sort of "Dead end" in the model, or perhaps an issue with how Chainner (is there another tool I can use to run the model on this image?) is implementing the upscaling of the image. Here's a zip file with the model and two frames:

https://www.mediafire.com/file/xxsv5yi6unr01bd/black_frame_omnisr.zip/file

Frame 1262 upscales correctly. 1263 produces a black image. I'm using Chainner 0.24.1.

One thing I did notice is that one of my manually created training images (white text on a black background designed to improve scaling of credits text) has some large flat black areas in it. I inserted noise into that area and updated the image pair in the training data, and I'm continuing training now. Not sure it'll make a difference, but it seems worth trying.

**Comment by neosr-project** [comment]
Hi @pokepress. Last time I checked chaiNNer had the same code for omnisr that neosr has, so that is probably not the issue.
One way of testing is by using `test.py` inside neosr to do the inference and compare the results. Just get [this template](https://github.com/neosr-project/neosr/blob/master/options/test_omnisr.toml) and modify the model path and scale. Then run `python test.py -opt test_omnisr.toml`. Results will be saved on the `experiments/` folder.

One possibility is that chaiNNer is forcing fp16 precision and that somehow causes these seemly random issues. If that is the case you could try disabling fp16 in chaiNNer.
Another possibility of course is that this is training related. If that is the case, then there are multiple things that could be the cause, such as your dataset, or some loss is interfering somehow. I recommend you use the [default configs](https://github.com/neosr-project/neosr/blob/master/options/train_omnisr.toml) before tweaking anything, just to make sure it's not some config issue.

**Comment by pokepress** [comment]
I haven't had a chance to try everything you suggested, but I did make my training config look more like the default omnisr one in terms of losses and whatnot. At this point I have a few clips from various episodes that tend to be problematic, and after a bit more training, there was one black frame across the four of them.

**Comment by pokepress** [comment]
Looks like this really is an FP16 issue. Both the test functionality and Chainner with FP16 disabled correctly upscale the affected frames. Presumably some sort of over/underflow is occurring and the upscale aborts for that frame. Usually FP16 is plenty for image processing, but perhaps certain values are causing issues.

**Comment by neosr-project** [comment]
Oh so it was fp16 :thinking: 
Interesting. Some networks are more prone to precision issues, omnisr is one of them. Good to know, now you can process it on chainner or other software without having to worry. Like I said, you could also use the `test.py` script, which runs in fp32 and also shouldn't cause any issue :+1:

**Comment by pokepress** [comment]
It is kind of a bummer, since fp16 is considerably faster. In theory since the black frames are sporadic I can re-run those scenes at fp32. But it's still extra work. Do you think it would be possible for Chainner to know when it hits a frame like this, and at the very least log it?

**Comment by neosr-project** [comment]
No idea, I'm not part of chaiNNer development. AFAIK, the fp16 onnx generated through the [`convert.py`](https://github.com/neosr-project/neosr/wiki/Model-Conversion) runs correctly without issues, so it might be something specific to the quantization being done by pytorch/chainner.
